testing/asserts.ts

```diff
@@ -647,12 +647,12 @@ export function assertThrows<E extends Error = Error>(
   ErrorClass?: new (...args: any[]) => E,
   msgIncludes?: string,
   msg?: string,
-): void;
+): Error | void;
```

Comment:
I think it would make sense to change this to returning `unknown` instead of `Error | void` since the thrown or rejected value could be a non string value. If called with an `ErrorClass` or `errorCallback`, I believe the return type should be `E` since it will assert that the thrown exception is of type `E`.

Reply:
I think it would make sense to update the catch block to also only do the following if an ErrorClass or errorCallback is set. That way these functions can be used for checking that non error values are thrown or rejected. If we did that, this would solve this issue too. https://github.com/denoland/deno_std/issues/1362#issuecomment-940101918

```ts
if (ErrorClass || errorCallback) {
  if (error instanceof Error === false) {
    throw new AssertionError("A non-Error object was thrown.");
  }
  assertIsError(
      error,
      ErrorClass,
      msgIncludes,
      msg,
  );
  if (typeof errorCallback == "function") {
      errorCallback(error);
  }
}
```

Reply:
Thank you for feedback. If understand correctly, we'd like to allow function throwing values not only of `Error` type. So the return type of function would change to `unknown` and it will look something like that:
```ts
function assertThrows<E extends Error = Error>(...): unknown {
  // ...
  catch (error) {
    if (ErrorClass || errorCallback) {
      if (error instanceof Error === false) {
        throw new AssertionError("A non-Error object was thrown.");
      }
      assertIsError(
        error,
        ErrorClass,
        msgIncludes,
        msg,
      );
      if (typeof errorCallback == "function") {
        errorCallback(error);
      }
    }
    err = error;
    doesThrow = true;
  }
  // ...
  return err;
}
```

Reply:
Yes, that's how I think it should work, then have return type be E for call signatures that include ErrorClass or errorCallback.

You can wait for a core team member to respond if you want. This is just my opinion as it would solve 2 issues at once while preserving backward compatibility. If someone wants to assert something was thrown other than an error, they would just do `const exception = assertThrows(() => { /* ... */ })` then make their assertions about exception.

Reply:
Right, makes sense for me. Let's wait for the input of the team core member.

Reply:
Here's what I meant in my last comment regarding call signatures and the reasoning behind them.

I think maybe something like this. The differences from the current call signature for the function are.
1. Added an `assertThrows` call signature that takes fn and optionally take msg.
2. Made `ErrorClass` not optional for the old first call signature.
3. Add a call signature without `ErrorClass` but with `msgIncludes`.
```ts
export function assertThrows(
  fn: () => unknown,
  msg?: string,
): unknown;
export function assertThrows<E extends Error = Error>(
  fn: () => unknown,
  // deno-lint-ignore no-explicit-any
  ErrorClass: new (...args: any[]) => E,
  msgIncludes?: string,
  msg?: string,
): E;
export function assertThrows(
  fn: () => unknown,
  msgIncludes: string,
  msg?: string,
): Error;
export function assertThrows(
  fn: () => unknown,
  errorCallback: (e: Error) => unknown,
  msg?: string,
): Error;
export function assertThrows<E extends Error = Error>(
  fn: () => unknown,
  errorClassOrCallback?:
    // deno-lint-ignore no-explicit-any
    | (new (...args: any[]) => E)
    | ((e: Error) => unknown),
  msgIncludesOrMsg?: string,
  msg?: string,
): E | Error | unknown {
  //...
}
```
With these call signatures, `assertThrows` could be used in the following ways.
```ts
// assert an exception was thrown. Type would be unknown.
const exception = assertThrows(() => { throw "oops" })
assertEquals(exception, "hello");
// asserts that the exception was an AssertionError that says oops. Returns it as type AssertionError so you can make additional assertions on it.
let error = assertThrows(() => { throw new AssertionError("oops") }, AssertionError, "oops")
// asserts that the exception was an Error of any kind that says oops. Returns it as type Error so you can make additional assertions on it.
error = assertThrows(() => { throw new AssertionError("oops") }, "oops")
// asserts that the exception was an Error of any kind that says oops. Returns it as type Error so you can make additional assertions on it.
error = assertThrows(() => { throw new AssertionError("oops") }, (error: Error) => {
  // here for backwards compatibility
  // it asserts the exception was an error before passing it to the callback.
})
```

We could also remove the last call signature where you can pass in an errorCallback since the same could be accomplished by doing the following then making your assertions on the returned value. I believe the only reason errorCallback was added was to work around the issue that it was previously decided that assertion function's cannot return values.

```ts
const error = assertThrows(() => { throw new Error("oops") }, Error);
// make additional assertions about error here instead of in an errorCallback.
```

We can let core team decide whether or not they want to remove the errorCallback. I'm personally think it should be removed.

Reply:
@KyleJune @mrkldshv 
I agree with removing errorCallback signature in the future, but let's first deprecate it. I think we can deprecate only specific overloaded signature by putting `@deprecated` jsdoc tag right above the deprecated one.

Suggested signature looks great to me 👍

Reply:
Alright, thanks for the feedback. I'll implement necessary changes 👍